lw/base/concepts.h

```diff
@@ -0,0 +1,32 @@
+#pragma once
+
+#include <concepts>
+#include <iterator>
+
+namespace lw {
+
+template <typename T, typename V>
+concept IterableAs =
```

Comment:
What's a concept?

Reply:
In C++20, you can define [type constraints using `concept`s][1]. These type constraints can then be used when defining templates (C++ version of generics). This allows you to write overloaded templates that are semi-specialized. You can see an example of this further on in this PR in the `Serializer::write` overloads where I use these concepts to direct how the serialization happens.

[1]: https://en.cppreference.com/w/cpp/language/constraints